File: Algoritms/Python/Sorting/NLogN/MergeSort.py
class MergeSort:
        def _merge(self, array, start, middle, end):
            # one array, but it's view as two arrays
            copy = array[::]

            indexA = start
            indexB = middle + 1

            indexK = start

            while indexA <= middle and indexB <= end:
                if copy[indexA] < copy[indexB]:
                    array[indexK] = copy[indexA]
                    indexA += 1
                else:
                    array[indexK] = copy[indexB]
                    indexB += 1

                indexK += 1

            while indexA <= middle:
                array[indexK] = copy[indexA]
                indexK += 1
                indexA += 1

            while indexB <= end:
                array[indexK] = copy[indexB]
                indexK += 1
                indexB += 1

        def _sort(self, array, start, end):

            if start < end:
                middle = (start + end) // 2
                self._sort( array, start, middle )
                self._sort( array, middle + 1, end )
                self._merge( array, start, middle, end )

        def sort(self, array):
            self._sort( array, 0, len( array ) - 1 )

File: Algoritms/Java/Sorting/NLogN/MergeSort.java
package NLogN;

import java.util.Arrays;

public class MergeSort<T extends Comparable<T>> {

	private void Merge(T[] array, int start, int middle, int end) {
		int arrayLength = array.length;
		T[] copy = Arrays.copyOf(array, arrayLength);

		int indexA = start;
		int indexB = middle + 1;

		int indexK = start;

		while (indexA <= middle && indexB <= end) {
			if (copy[indexA].compareTo(copy[indexB]) < 0) {
				array[indexK++] = copy[indexA++];
			} else {
				array[indexK++] = copy[indexB++];
			}

		}

		while (indexA <= middle) {
			array[indexK++] = copy[indexA++];
		}

		while (indexB <= end) {
			array[indexK++] = copy[indexB++];
		}
	}

	public void Sort(T[] array) {
		this.Sort(array, 0, array.length - 1);
	}

	private void Sort(T[] array, int start, int end) {
		if (start < end) {
			int middle = (start + end) / 2;

			this.Sort(array, start, middle);
			this.Sort(array, middle + 1, end);

			this.Merge(array, start, middle, end);

		}
	}
}
